### PR TITLE
Faker: Emit correct error on exit when running without GLX

### PR DESCRIPTION
**Pull request by aristotelos**
Commit bdc83ac67d7df6c7c62a3ea96af7ad0d730f5f92 introduced emitting GLX
error instead of throwing exceptions. However, when GLX is not enabled,
the faker will exit writing "The EGL back end requires a 2D X server
with a GLX extension." to stderr on any error. This hides the original
error and is not a valid way to indicate what failed.

This fix emits error details instead of erroneously reporting that GLX
should be enabled.

### CODE REVIEW

**Comment by dcommander** [comment]
That error message doesn't "hide" the original error.  It ***is*** the original error.  Because the EGL back end doesn't have a 3D X server on which to hang GLX error handling, it has to hang GLX error handling on the 2D X server.  If the 2D X server lacks a GLX extension, then the EGL back end can't do that.  In fact, your patch hides some critical information from the user, which is that VirtualGL needs a GLX extension (on the 3D X server when using the GLX back end and on the 2D X server when using the EGL back end) in order to properly function.  The current error message is both informative and instructive.  It tells the user what they must do (run VirtualGL on a 2D X server with a GLX extension) in order for GLX error handling to work properly.  Once that is done, the faker will throw the correct GLX error.  Your patch obscures that requirement by pretending that GLX error handling is working properly, when in fact it isn't.  You are essentially implementing a shadow GLX error handler that unnecessary reinvents the actual GLX error handler, when the correct thing to do is inform the user that the GLX error handler isn't working because there is no GLX extension.

**Comment by dcommander** [comment]
Also, the other patch you submitted with this PR is specific to your downstream build preferences and does not belong upstream.  Your personal choice of build directories should be excluded in **.git/info/exclude**, not in **.gitignore**.  (You can maintain your own CMake preferences in your own build tree and exclude them in **.git/info/exclude** as well.) 
 **.gitignore** is for things that affect everyone, and not everyone uses **out** as a build directory name.  (I personally don't.)

**Comment by aristotelos** [comment]
> That error message doesn't "hide" the original error. It _**is**_ the original error. Because the EGL back end doesn't have a 3D X server on which to hang GLX error handling, it has to hang GLX error handling on the 2D X server. If the 2D X server lacks a GLX extension, then the EGL back end can't do that. In fact, your patch hides some critical information from the user, which is that VirtualGL needs a GLX extension (on the 3D X server when using the GLX back end and on the 2D X server when using the EGL back end) in order to properly function. The current error message is both informative and instructive. It tells the user what they must do (run VirtualGL on a 2D X server with a GLX extension) in order for GLX error handling to work properly. Once that is done, the faker will throw the correct GLX error. Your patch obscures that requirement by pretending that GLX error handling is working properly, when in fact it isn't. You are essentially implementing a shadow GLX error handler that unnecessary reinvents the actual GLX error handler, when the correct thing to do is inform the user that the GLX error handler isn't working because there is no GLX extension.

We have been using VirtualGL without the GLX extension for a long time already... We use it to redirect to a headless EGL NVIDIA context to be able to run VNC server which doesn't support OpenGL. So you say this is not a supported use case?

**Comment by dcommander** [comment]
> We have been using VirtualGL without the GLX extension for a long time already... We use it to redirect to a headless EGL NVIDIA context to be able to run VNC server which doesn't support OpenGL. So you say this is not a supported use case?

Correct.  I don't know what you mean by a "long time", but the EGL back end was introduced 3 years ago in VirtualGL 3.0 beta1 and stabilized 2 1/2 years ago in VirtualGL 3.0.  The change log for VirtualGL 3.0 beta1 states that the EGL back end requires a 2D X server with a GLX extension (although, to be clear, the GLX extension can be software-only, since the EGL back end only uses it for error handling.)  TurboVNC has included a GLX extension since 2.2.x, which was released more than 6 years ago, and [no prior versions of TurboVNC are still supported](https://turbovnc.org/DeveloperInfo/Versioning).  TigerVNC has included a GLX extension all along, and as far as I remember, RealVNC has included a GLX extension since v4.x.  I am not aware of any actively-supported Xvnc implementations that lack a GLX extension.  (TightVNC 1.3.x, as an example, did not include a GLX extension, but TightVNC 1.3.x is no longer supported.)  Also, any Xvnc implementation other than TurboVNC or TigerVNC will have suboptimal performance with VirtualGL.  TurboVNC is our project's Xvnc implementation and is the implementation we recommend and support for use with VirtualGL.

If the 2D X server lacks a GLX extension, then the EGL back end will print a warning if the GLX application calls `XQueryExtension(..., "GLX", ...)`.  The application will continue to work, but if it encounters a GLX error, then VirtualGL will throw the aforementioned fatal error to indicate that GLX error handling is broken.

VirtualGL is a GLX implementation, so it has to conform to the GLX spec.  That means throwing specific GLX errors, using the X11 error handler, in response to specific conditions in specific GLX functions.  (fakerut, the VirtualGL Faker unit test, tests whether the Faker's GLX error handling is conformant with the spec.  I also use piglit to test GLX conformance, and it also tests whether certain GLX errors are handled correctly.)  When I was developing the EGL back end, I could not figure out how to conformantly and robustly implement GLX error handling without a GLX extension.  Since all of the supported 2D X servers at the time had a GLX extension, it made sense to require that rather than to sacrifice conformance. Like I said, if you're using the GLX back end in VirtualGL, then the 2D X server does not need a GLX extension (because GLX error handling piggybacks off of the 3D X server in that case.)  That was the "traditional" way of running VirtualGL.  "Back in the day", TurboVNC didn't need a GLX extension for that reason and because window managers didn't use OpenGL.  However, the EGL back end is a much more recent development, and it is somewhat legacy-free in the sense that it doesn't support old X servers without a GLX extension, nor does it support obsolete OpenGL features.

**Comment by tsondergaard** [comment]
Hi @dcommander, we are using a very old and heavily modified copy of RealVNC and it does not have a GLX extension. We have been using virtualGL with the EGL backend since around March 2022 and we have just accepted that if VirtualGL runs into a situation where it wants to report an error then it will just fail hard because of the lack of GLX extension in our old X11 server. The intention with @aristotelos PR is to include a bit more information in the error message so you can see not only that the faker exited due to the missing GLX extension, but also see the root error that the faker wanted to report to the application.

As for the CMakePresets.json being added, it's fine that you don't want it. More and more projects add them as a convenience. Adding them does not mean that everyone has to use them.

**Comment by dcommander** [comment]
> Hi @dcommander, we are using a very old and heavily modified copy of RealVNC and it does not have a GLX extension. We have been using virtualGL with the EGL backend since around March 2022 and we have just accepted that if VirtualGL runs into a situation where it wants to report an error then it will just fail hard because of the lack of GLX extension in our old X11 server. The intention with @aristotelos PR is to include a bit more information in the error message so you can see not only that the faker exited due to the missing GLX extension, but also see the root error that the faker wanted to report to the application.

I understand, but as with any software, VirtualGL has a scope of support.  That is especially necessary because I am an independent developer who has to do everything-- software development, testing, marketing, business development, project management, product management, accounting, IT, tech support, etc.-- myself.  Thus, I have a strong bias against enhancements that make it more difficult to support the software, unless they fundamentally improve the experience for most users.

The PR, as written, is only useful for a configuration that the EGL back end doesn't support and has never supported. Furthermore, the PR proclaims that your desired downstream behavior is "correct", which implies that our existing (and documented) upstream behavior is incorrect.  It also introduces potentially hard-to-maintain and error-prone code that reinvents an existing subsystem (the GLX error handler.)  It also eliminates the existing error message, which provided useful information to the user regarding the fact that they are exceeding VirtualGL's scope of support. Our scope of support is generous and doesn't limit you to using our X proxy (TurboVNC), but it does limit you to using an X proxy with a GLX extension if you want to use the EGL back end.

I would be more receptive to an enhancement that frees the EGL back end from the GLX extension requirement, but such an enhancement would have to be maintainable and conformant with the GLX spec.  Again, despite my expertise with this code, I was not able to devise such a solution.  The only conformant solution I was able to devise was to use the 2D X server's GLX extension for error handling, which is why a GLX extension is a documented requirement of the EGL back end.  As of March 2022, when you started using the EGL back end, that requirement had been documented for about 18 months.

Technically speaking, I could change `faker::sendGLXError()` into a macro that prints the GLX error information by stringifying the arguments, thus making it possible to do what you want without creating a maintenance burden for me.  However, that would only provide the illusion of support.  Let's be clear about the fact that, when using the EGL back end, `XQueryExtension(... "GLX", ...)` and `glXQueryExtension()` will return `False` if the 2D X server doesn't have a GLX extension.  That means that it is not legal for calling applications to call GLX functions after that point.  The fact that your configuration works at all with the EGL back end is purely an accident.  It definitely would not work with applications that check for the existence of a GLX extension before using it.

You have several options:
1. Rather than continuing to use an X proxy that is neither supported by VirtualGL nor RealVNC, work with me to introduce any features into TurboVNC that your custom RealVNC implementation has but TurboVNC lacks.  Not only would this bring you into VirtualGL's scope of support, but it would likely improve the performance of your solution as well.
2. Continue doing what you're doing but maintain a downstream fork of VirtualGL with patches, such as this one, that are applicable only to configurations that VirtualGL doesn't support.  In other words, if you insist on using a modified version of RealVNC that exceeds our scope of support, then it is incumbent upon you to use a modified version of VirtualGL as well.
3. Work with me on a solution to free the EGL back end from the GLX extension requirement. (Again, however, I was not able to find such a solution myself.  I don't claim to be all-knowing, so perhaps there is something I missed, but I am skeptical of that.)

> As for the CMakePresets.json being added, it's fine that you don't want it. More and more projects add them as a convenience. Adding them does not mean that everyone has to use them.

It isn't appropriate for this project, because we have different recommended build configurations depending on the platform.  Also, per above, the CMakePresets patch assumed a particular build directory nomenclature and location, which definitely isn't appropriate for upstream inclusion in any project.  (The build directory nomenclature/location is absolutely a downstream preference.)  If I were so presumptuous as to include such build system defaults, I would at least make the defaults match my own preferences, not someone else's.  What I do instead is maintain a set of [official build scripts](https://github.com/VirtualGL/buildscripts) and [Docker images](https://github.com/VirtualGL/docker), which implement the build configurations that we use when building official VirtualGL packages but which do not bind users in any way to those configurations.  

In summary, both of these patches have basically the same problem, which is that they ask me to adopt upstream behavior that is only really suitable for your specific downstream configuration.
